---
title: "Differential expression analysis of adult and late third instar larval testes from Drosophila melanogaster males and translocationsT(1;2) males."
author: "Analysis done by Yuejun Wang, TriLab Biounformatic Group"
output:
  html_document:
  pdf_document: default
  df_print: paged
gemoetry: margin=1in
fontsize: 8pt
fontfamily: mathpazo
---

\center

# *Summary:*
In order to study the effect of translocation T(1;2) on gene expression in testes, we performed polyA+ RNA-seq profiles from pools of adult and third instar larval testes from 7 translocation lines in biological quadruplicate. Among the 7 translocations, 5 are male sterile, 2 are fertile. In order to compare the expression profiles of translocations to the control, we sequence the testes from the parental flies that were irradiated to generate the translocation lines. We prepared 50bp stranded single end read libraries.

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, tidy = TRUE)
```

# Load libraries
```{r message=FALSE}

options(java.parameters = "-Xmx8000m")

library(tidyverse)
library(DESeq2)
library(cowplot)
library(ggpubr)
library(RColorBrewer)
library(pheatmap)
library(ggsci)
library(AnnotationDbi)
library(scales)
library(org.Dm.eg.db)
library(xlsx)
library(r2excel)
library(ggplot2)
library(viridis)
library(RSkittleBrewer)
library(genefilter)
library(corrplot)
library(reshape2)
library(ComplexHeatmap)
library(circlize)
library(UpSetR)
library(zoo)
library(karyoploteR)
library(ggthemes)
trop = RSkittleBrewer('tropical')
```

# Load aux functions
```{r}
source(file = "../01_aux_rnaseq_functions.R")
```

# Load data
```{r}
metadata <- as.data.frame(read_tsv(file = "../SampleInfo.txt", col_names = TRUE, comment = "#"))
rownames(metadata) <- metadata$Sample


read_counts <- as.data.frame(read_tsv(file = "/gpfs/gsfs12/users/wangy80/TK105/results/05counts/read_counts", col_names = TRUE, comment = "#"))
rownames(read_counts) <- read_counts$Geneid # adding gene ids as row names


# Rename sample names
read_counts <- read_counts[,7:length(read_counts)]

colnames(read_counts) <- gsub(".sorted.dedup.bam","",
                              gsub("results/04dedup/","",
                                   colnames(read_counts)))


# Sort tables so metadata and read counts match order
read_counts <- read_counts[,match(metadata$SampleID, colnames(read_counts))]
all(metadata$SampleID == colnames(read_counts))


# Round read counts to the closest interger
read_counts <- round(read_counts, digits = 0)

# include total read counts in metadata
metadata$read_counts <- colSums(read_counts)

# Rename column names in read_counts based on metadata
colnames(read_counts) <- rownames(metadata)

write.table(x = metadata, file = "metadata.txt", sep = "\t") 
```

# DE analysis with DESeq2

```{r}
dir.create(path = "./Plots", showWarnings = FALSE)

# Converting Time, DMSO and Replicate number to factors
metadata$Age <- as.factor(metadata$Age)
metadata$Condition <- relevel(as.factor(metadata$Condition),"Control")
metadata$Replicate <- as.factor(metadata$Replicate)

# Adding read_depth in design to control for read_depth
dds <- DESeqDataSetFromMatrix(countData = read_counts, 
                              colData = metadata,  
                              design = ~ Age + Condition + Age:Condition)


# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "Sample", 
          y = "read_counts",
          x.text.angle = 90,
          fill = "Condition", 
          title = "Total read counts per sample", 
          ylab = "Read counts",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc")

ggsave2("Plots/barplot_read_counts_per_sample.pdf", plot = p,width=14,height=8)
p
# Normalize counts
vsd <- vst(dds, blind=FALSE)

# Keep genes with at least 20 reads total across samples
#keep <- rowSums(as.data.frame(dds@assays@data@listData)) >= 20

keep <- rowSums(counts(dds) >= 10) >= 3
vsd <- vsd[keep,]

# Calculate distances between samples
sampleDists <- dist(t(assay(vsd)))

# Plot inter-sample distances
old.par <- par(no.readonly=T)

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Sample)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
p.hm <- pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         color = colorRampPalette(c("red", "white", "blue"))(100))

pdf("./Plots/heat_map.pdf", p.hm,width=16,height=12)
p.hm
dev.off()

p.hm

# PCA
pcaData <- plotPCA(vsd, intgroup=c("Condition","Age"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords
p1 <- ggplot(pcaData, aes(PC1, PC2, color=Condition, shape=Age)) +
  geom_point() + scale_color_lancet() + 
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) 

ggsave("Plots/pca.pdf", plot = p1)
p1

```

Samples clustered mainly by Age.

# filter outliners

```{r}
dds.f <- dds[,dds$Sample != "Adult_Testes_BM24_Sterile_translocation_rep3" & dds$Sample != "Adult_Testes_BM24_Sterile_translocation_rep2" & dds$Sample != "Adult_Testes_BM6_Sterile_translocation_rep3"]

vsd.f <- vst(dds.f, blind=FALSE)

df <- as.data.frame(assay(vsd.f))
write.table(df,"DESeq2_normalized.txt",sep="\t",quote=F)


# PCA
pcaData <- plotPCA(vsd.f, intgroup=c("Condition","Age"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords
p1 <- ggplot(pcaData, aes(PC1, PC2, color=Condition, shape=Age)) +
  geom_point() + scale_color_lancet() + 
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) 

ggsave("Plots/pca.f.pdf", plot = p1)
p1


```

# Separate L3 and Adult

```{r}
dds.Adult <- dds.f[, dds.f$Age == "Adult"]
dds.Adult$Age <- droplevels(dds.Adult$Age)

design(dds.Adult) <- ~ Condition

dds.Adult <- DESeq(dds.Adult)
dds.Adult$dds_ID = "Adult"
resultsNames(dds.Adult)


dds.L3 <- dds.f[, dds.f$Age == "L3"]
dds.L3$Age <- droplevels(dds.L3$Age)

design(dds.L3) <- ~ Condition

dds.L3 <- DESeq(dds.L3)
dds.L3$dds_ID = "L3"
resultsNames(dds.L3)
```


# Perform Age-specific PCA analysis

```{r}
# Adult
# Normalize counts
vsd.Adult <- vst(dds.Adult, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(counts(dds.Adult) >= 10) >= 3

vsd.Adult <- vsd.Adult[keep,]

# Calculate distances between samples
sampleDists.Adult <- dist(t(assay(vsd.Adult)))

# Define the number of colors you want
colourCount = length(unique(vsd.Adult$Condition))
getPalette = colorRampPalette(brewer.pal(12, "Paired"))

# PCA
pcaData.Adult <- DESeq2::plotPCA(vsd.Adult, intgroup=c("Condition"), returnData=TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData.Adult, "percentVar"))
y.coords = c(min(pcaData.Adult$PC1, pcaData.Adult$PC2), max(pcaData.Adult$PC1, pcaData.Adult$PC2))
x.coords = y.coords
p1 <- ggplot(pcaData.Adult, aes(PC1, PC2, color=Condition)) +
  geom_point(size=4) + 
  scale_color_manual(values = getPalette(colourCount)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData.Adult$PC1)-min(pcaData.Adult$PC1))/(max(pcaData.Adult$PC2)-min(pcaData.Adult$PC2))) + labs(title = "Adult")

ggsave("Plots/pca_by_Adult.pdf", plot = p1)
p1

## L3 ## 
# Normalize counts
vsd.L3 <- vst(dds.L3, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(counts(dds.L3) >= 10) >= 3

vsd.L3 <- vsd.L3[keep,]

# Calculate distances between samples
sampleDists.L3 <- dist(t(assay(vsd.L3)))

# Define the number of colors you want
colourCount = length(unique(vsd.L3$Condition))
getPalette = colorRampPalette(brewer.pal(12, "Paired"))

# PCA
pcaData.L3 <- DESeq2::plotPCA(vsd.L3, intgroup=c("Condition"), returnData=TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData.L3, "percentVar"))
y.coords = c(min(pcaData.L3$PC1, pcaData.L3$PC2), max(pcaData.L3$PC1, pcaData.L3$PC2))
x.coords = y.coords
p2 <- ggplot(pcaData.L3, aes(PC1, PC2, color=Condition)) +
  geom_point(size=4) + 
  scale_color_manual(values = getPalette(colourCount)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData.L3$PC1)-min(pcaData.L3$PC1))/(max(pcaData.L3$PC2)-min(pcaData.L3$PC2))) + labs(title = "L3")

ggsave("Plots/pca_by_L3.pdf", plot = p2)
p2
```


# Run DE analysis
```{r}

dir.create(path = "./DE", showWarnings = FALSE)

comparisons <- list(
c("Condition_BM1_Sterile_vs_Control"),
c("Condition_BM21_Fertile_vs_Control"),
c("Condition_BM24_Sterile_vs_Control"),
c("Condition_BM4_Fertile_vs_Control"),
c("Condition_BM46_Sterile_vs_Control"),
c("Condition_BM5_Sterile_vs_Control"),
c("Condition_BM6_Sterile_vs_Control")
)


########################################################################################################
# Make function here to compute DE analysis and output results for each comparison above.
########################################################################################################
# Define function for processing and saving result tables
sort_and_write_res_table <- function(result_table, 
                                     file_name = "TABLE.xlsx", 
                                     sheet_name = "sheet_1"){
  dir.create(path = "./DE", showWarnings = FALSE)
  
  # add path
  file_name <- paste0("./DE/",file_name)
  
  # Check if an excel spreadsheet already exists, otherwise create one
  if(file.exists(file_name)){
    wb <- loadWorkbook(file = file_name)
  } else {
    wb <- createWorkbook(type="xlsx")
  }
  
  # # Create new excel sheet, remove sheets if it already exits (if the sheet name is too long, the errors might occur)
  sheets <- getSheets(wb)
  if(is.element(sheet_name,names(sheets))){
    removeSheet(wb, sheetName=sheet_name)
  }
  sheet <- createSheet(wb, sheetName = sheet_name)

  # Sort genes by (padj)
  result_table_sorted <- result_table[order(result_table$padj, decreasing = FALSE),]
  
  # Add gene symbols
  symbol_list <- replace_gene_acc_by_symbol_ids(rownames(result_table_sorted), return_all = TRUE, db = org.Dm.eg.db)
  result_table_sorted$Gene_name <- symbol_list
  result_table_sorted <- as.data.frame(result_table_sorted) 
  
  # Add df to excel sheet
  xlsx.addTable(wb = wb, sheet = sheet, data = result_table_sorted, startRow = 1, startCol = 1)

  # Write sorted table to Excel file as different worksheets. Need file name + Worksheet name !!!
  saveWorkbook(wb, file_name)
  
  # Write sorted table to file
  # write.table(as.data.frame(result_table_sorted), file = paste0("./DE/",file_name,".txt"), sep = "\t", col.names=NA)
  
  return(result_table_sorted)
}


get_deseq_result <- function(dds, contrast, analysis_type = "ashr"){
  #print(1)
  if(analysis_type %in% c("ashr","apeglm","normal")){
    res <- lfcShrink(dds = dds, contrast =list(contrast), type = analysis_type)
  } else if (analysis_type == "result"){
    res <- results(object = dds, contrast = contrast, independentFiltering = FALSE)
  }
  #print(2)
  # Replace NAs by 1s
  res$pvalue[is.na(res$pvalue)] <- 1
  res$padj[is.na(res$padj)] <- 1
  #print(3)
  # Print out summary of results
  print(contrast)
  summary(res, alpha = 0.05)
  #print(4)
  # Sort result table and save it
  my_file_name = paste0("DE_",dds$dds_ID[1])
  #paste0("DE_",paste0(contrast, collapse = "_"),"_",dds$dds_ID[1])
  
  my_sheet = paste0(paste0(contrast, collapse = "_"),"_",dds$dds_ID[1])
  my_sheet = gsub("Condition_","C_",my_sheet)
  my_sheet = gsub("Sterile","S",my_sheet)
  my_sheet = gsub("Fertile","F",my_sheet)
  #print(5)
  res_sorted <- sort_and_write_res_table(result_table = res, 
                                         file_name =  paste0(my_file_name,".xlsx"), 
                                         sheet_name = my_sheet)
  #print(6)
  return(res_sorted)
  
}
```

```{r}
# Using lfcShrink instead of results to reduce high Log2FC bias of genes with low expression

DE_results.Adult = list()
DE_results.L3 = list()
for (i in 1:length(comparisons)){
  print(i)
  id <- paste0(comparisons[[i]], collapse = "_")
  DE_results.Adult[[id]] <- get_deseq_result(dds = dds.Adult, contrast = comparisons[[i]], analysis_type = "ashr" )
  DE_results.L3[[id]] <- get_deseq_result(dds = dds.L3, contrast = comparisons[[i]], analysis_type = "ashr" )

  genes_of_interest.Adult <- head(rownames(DE_results.Adult[[id]][DE_results.Adult[[id]]$padj <= 0.05, ]))
  if (length(genes_of_interest.Adult) > 0){
    plot_normalized_gene_expression(my_dds = dds.Adult, 
                                  ensmbl_gene_list = genes_of_interest.Adult, 
                                  file_prefix = paste0("Gene_exp_",id, "_Adult_")) 
  }
  genes_of_interest.L3 <- head(rownames(DE_results.L3[[id]][DE_results.L3[[id]]$padj <= 0.05, ]))
  if (length(genes_of_interest.L3) > 0){
    plot_normalized_gene_expression(my_dds = dds.L3, 
                                  ensmbl_gene_list = genes_of_interest.L3, 
                                  file_prefix = paste0("Gene_exp_",id, "_L3_")) 
  }
}  

```

# Build summary table across contrats

```{r}
N <- length(names(DE_results.Adult))
# Initialize table
my_table <- data.frame(rbind("logFC.up.Adult" = rep(0, N), 
                             "logFC.down.Adult" = rep(0, N),
                             "logFC.up.L3" = rep(0, N), 
                             "logFC.down.L3" = rep(0, N)
                             )
                       ) 
colnames(my_table) <- names(DE_results.Adult)

for (i in names(DE_results.Adult)){ 
  my_table["logFC.up.Adult",i] <- table(DE_results.Adult[[i]]$padj <= 0.05 & DE_results.Adult[[i]]$log2FoldChange > 0)[2]
  my_table["logFC.down.Adult",i] <- table(DE_results.Adult[[i]]$padj <= 0.05 & DE_results.Adult[[i]]$log2FoldChange < 0)[2]
  my_table["logFC.up.L3",i] <- table(DE_results.L3[[i]]$padj <= 0.05 & DE_results.L3[[i]]$log2FoldChange > 0)[2]
  my_table["logFC.down.L3",i] <- table(DE_results.L3[[i]]$padj <= 0.05 & DE_results.L3[[i]]$log2FoldChange < 0)[2]
}

names(DE_results.Adult) <- paste0("Adult_",names(DE_results.Adult))
names(DE_results.L3) <- paste0("L3_",names(DE_results.L3))


# Replace NA by 0s
my_table[is.na(my_table )] <- 0
write.table(x = my_table, file = "DE_summary.txt", sep = "\t", col.names = NA)

```


# Look for DE gene overlaps among time points and direction of change in gene expression
# Build summary table across contrats
```{r}

get_diff_exp_genes_across_conditions_plot <- function(DE_result_list){
  # Initialize list 
  my_overlap_list.f <- list()
  all_sig_genes <- c()
  
  for (i in names(DE_result_list)){
    genes_up.f <- c(rownames(subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange > 0.5)))
    genes_dn.f <- c(rownames(subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange < -0.5)))
    #my_overlap_list.f[i] <- c(subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange > 0)[,"Gene_name"])
    my_overlap_list.f[[i]] <- unique(c(genes_up.f, genes_dn.f) )
    #                          subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange < 0)[,"Gene_name"]
    #                          )
    all_sig_genes <- unique(c(all_sig_genes, my_overlap_list.f[[i]]))
    #my_table_overlap["logFC.up.male",i] <- subset(DE_results.m[[i]], padj <= 0.05 & log2FoldChange > 0)[,"Gene_name"]
    #my_table_overlap["logFC.down.male",i] <- subset(DE_results.m[[i]], padj <= 0.05 & log2FoldChange < 0)[,"Gene_name"]
  }
  
  row.N <- length(all_sig_genes)
  col.N <- length(names(DE_result_list))
  my_tibble <- as_tibble(matrix(nrow = row.N, ncol = col.N + 2))
  colnames(my_tibble) <- c("gene_ids","gene_names", names(DE_result_list))
  my_tibble$gene_ids <- all_sig_genes
  my_tibble$gene_names <- subset(DE_result_list[[i]], rownames(DE_result_list[[i]]) %in%  all_sig_genes)["Gene_name"]
  
  # Load tibble (1 => is differentially expressed in contrast X; 0 => is NOT differentially expressed in contrast X)
  for (my_row in 1:row.N){
    for (my_col in 1:col.N){
      my_query <- my_tibble$gene_ids[my_row]
      my_contrast <- colnames(my_tibble)[my_col]
      my_tibble[my_row, my_col + 2] <- ifelse(test = my_query %in% my_overlap_list.f[[my_contrast]], 
                                          yes = subset(DE_result_list[[my_contrast]], rownames(DE_result_list[[my_contrast]]) == my_query)["log2FoldChange"], 
                                          no = 0)
    }
  }
  
  my_tibble_long <- pivot_longer(my_tibble, cols = 3:ncol(my_tibble), names_to = "contrast", values_to = "log2fc")
  my_tibble_long$log2fc[my_tibble_long$log2fc < -5] <- -5
  my_tibble_long$log2fc[my_tibble_long$log2fc >  5] <-  5
  
  # Plot table
  p <- my_tibble_long %>% ggplot(aes(x = gene_names$Gene_name, y = contrast, colour = "white", fill = log2fc)) + 
    theme_pubclean() + 
    geom_tile( ) +
    scale_fill_gradient2(low = "red", high = "green", midpoint = 0, mid = "white") + 
    coord_fixed() + 
    theme(axis.ticks = element_blank(),
          panel.grid.minor.x = element_blank(),panel.grid.minor.y = element_blank(),
          panel.grid.major.x = element_blank(),panel.grid.major.y = element_blank(),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + 
    xlab("Differentially expressed genes") + 
    ylab("Contrasts") + 
    labs(title = "Log2FC of genes differentially expressed across experimental conditions", 
         subtitle = "( | Log2FC | > 0.5 and adj.p < 0.05 )")

    return(p)
}

lists <- c(DE_results.Adult,DE_results.L3)

p <- get_diff_exp_genes_across_conditions_plot(DE_result_list = lists)
ggsave(filename = "perturbed_genes.pdf", plot = p, path = "./Plots", width = 20, height = 10)
p

p.f <- get_diff_exp_genes_across_conditions_plot(DE_result_list = DE_results.Adult)
ggsave(filename = "perturbed_genes_Adult.pdf", plot = p.f, path = "./Plots", width = 20, height = 10)
p.f

p.m <- get_diff_exp_genes_across_conditions_plot(DE_result_list = DE_results.L3)
ggsave(filename = "perturbed_genes_L3.pdf", plot = p.m, path = "./Plots", width = 20, height = 10)
p.m
```

## heatmap for all comparisons on all significant genes detected in at least one comparisons

```{r message=FALSE}

heatmap_DE <- function(list=list,out=out) {
  list<-lapply(list, function(x) 
  cbind(x, gene_ids = rownames(x)))

  filter <- function(df) {
    df[df$padj<=0.05 & abs(df$log2FoldChange)>0.5,]
  }

  # Apply the filter function to each data frame in the list
  filtered <- lapply(list, filter)

  # Merge the filtered data frames into a single data frame
  all_sig_genes <- unique((bind_rows(filtered))$gene_ids)

  # Merge data frames by gene_ids
  merged_df <- bind_rows(list, .id = "comparisons")

  filtered_df <- merged_df[merged_df$gene_ids %in% all_sig_genes, ]

  filtered_df_wide <- pivot_wider(filtered_df[,c(1,3,8)],  names_from = "comparisons", values_from = "log2FoldChange")


  # heatmap

  mat <- as.data.frame(filtered_df_wide[,2:ncol(filtered_df_wide)])
  rownames(mat) <- filtered_df_wide$gene_ids

  mat[is.na(mat)] <- 0
  write.table(mat,paste0(out,"_log2FC0.5_padj0.05.txt"),sep="\t",quote=F,row.names=T)
  
  
  p <- Heatmap(mat, 
        name="log2FoldChange",
        show_row_names = FALSE,
        col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))
        #column_names_gp = gpar(fontsize =8),
        #heatmap_height = unit(12, "cm"), 
        #heatmap_width = unit(12, "cm")
        )

  pdf(paste0("./Plots/pheatmap_",out,".pdf"),height=12,width=12)
  plot(p)
  dev.off()
  return(mat)
}

mat_Adult = heatmap_DE(list=DE_results.Adult,out="Adult_DE")
mat_L3 = heatmap_DE(list=DE_results.L3,out="L3_DE")
mat_all = heatmap_DE(list=lists,out="all_DE")

```

## rank by log2FC

```{r}

annotation <- read.delim("/gpfs/gsfs12/users/wangy80/TK105/data/00ref/dmel-all-r6.46.gene.anno.txt",header=F,sep="\t")
names(annotation) <- c("chr","start","end","strand","GeneID","Gene_name")
annotation <- annotation[which(annotation$chr=="2L" | annotation$chr=="2R" | annotation$chr=="3L" | annotation$chr=="3R" | annotation$chr=="4" | annotation$chr=="X" | annotation$chr=="Y" | annotation$chr=="mitochondrion_genome"),]


window_size <- 30
ymin=-0.5
ymax=0.5


# window_size <- 20,30
# ymin=-1
# ymax=1

for(i in names(lists)){
  df <- as.data.frame(lists[i])
  names(df) <- c("baseMean","log2FoldChange","lfcSE","pvalue","padj","Gene_name")
  res <- inner_join(df,annotation,by="Gene_name")
  
  res$log2FoldChange <- ifelse(res$padj<0.05,res$log2FoldChange,0)
  # Sort genes by location
  sorted_genes <- res %>% arrange(chr, start)

  # Define window size

  # Calculate sliding window averages for each chromosome
  sliding_window_averages <- sorted_genes %>%
    group_by(chr) %>%
    mutate(window_average = rollapply(log2FoldChange, window_size, mean, align = "left", fill = NA)) %>%
    filter(!is.na(window_average))


  pdf(paste0("./Plots/",i,"_line_kp_w",window_size,".pdf"),width = 20, height = 10)
  kp <- plotKaryotype(genome="dm6", plot.type=1, main=i)
  kpLines(kp, chr=paste0("chr",sliding_window_averages$chr), x=sliding_window_averages$start, y=sliding_window_averages$window_average,
          col="#AA88FF", ymin=ymin, ymax=ymax, r0=0, r1=1)
  kpAxis(kp, ymin=ymin, ymax=ymax, r0=0, r1=1, col="gray50", cex=1)
  kpAbline(kp, h=0.5, data.panel = 2, r0=0, r1=1, col="red")
  dev.off()
            
}

```

# combine all conditions

```{r}

# Initialize an empty data frame for the result
merged_df <- data.frame()

# Iterate over the list, merging each data frame and adding the source name
for (name in names(DE_results.L3)) {
  # Temporary data frame to modify and then bind
  temp_df <- DE_results.L3[[name]]
  temp_df$class <- name  # Create a new column with the name of the list
  
  # Bind this temp data frame to the merged_df
  merged_df <- rbind(merged_df, temp_df)
}

res <- inner_join(merged_df,annotation,by="Gene_name")
  
res$log2FoldChange <- ifelse(res$padj<0.05,res$log2FoldChange,0)
# Sort genes by location
sorted_genes <- res %>% arrange(class,chr, start)

# Define window size
window_size <- 50

# Calculate sliding window averages for each chromosome
sliding_window_averages <- sorted_genes %>%
  group_by(class,chr) %>%
  mutate(window_average = rollapply(log2FoldChange, window_size, mean, align = "left", fill = NA)) %>%
  filter(!is.na(window_average))

#Plot the sliding window averages
p <- ggplot(sliding_window_averages, aes(x = start, y = window_average,color=class)) +
  facet_wrap(~chr,ncol=1) +
  geom_line(alpha=1) +
  labs(x = "Location", y = "Average Log2FoldChange", title = "Sliding Window log2FoldChange by Chromosome") +
  scale_color_brewer(palette="Paired") + 
  ylim(c(-0.5,0.5)) +
  geom_hline(yintercept=0,color="#999999",linetype=2) +
  theme_few()

ggsave(filename = paste0("L3_line_w",window_size,".pdf"), plot = p, path = "./Plots", width = 20, height = 10)
  
p
```
```{r}

# Initialize an empty data frame for the result
merged_df <- data.frame()

# Iterate over the list, merging each data frame and adding the source name
for (name in names(DE_results.Adult)) {
  # Temporary data frame to modify and then bind
  temp_df <- DE_results.Adult[[name]]
  temp_df$class <- name  # Create a new column with the name of the list
  
  # Bind this temp data frame to the merged_df
  merged_df <- rbind(merged_df, temp_df)
}

res <- inner_join(merged_df,annotation,by="Gene_name")
  
res$log2FoldChange <- ifelse(res$padj<0.05,res$log2FoldChange,0)
# Sort genes by location
sorted_genes <- res %>% arrange(class,chr, start)

# Define window size
window_size <- 10

# Calculate sliding window averages for each chromosome
sliding_window_averages <- sorted_genes %>%
  group_by(class,chr) %>%
  mutate(window_average = rollapply(log2FoldChange, window_size, mean, align = "left", fill = NA)) %>%
  filter(!is.na(window_average))

#Plot the sliding window averages
p <- ggplot(sliding_window_averages, aes(x = start, y = window_average,color=class)) +
  facet_wrap(~chr,ncol=1) +
  geom_line(alpha=1) +
  labs(x = "Location", y = "Average Log2FoldChange", title = "Sliding Window log2FoldChange by Chromosome") +
  scale_color_brewer(palette="Paired") + 
  ylim(c(-0.5,0.5)) +
  geom_hline(yintercept=0,color="#999999",linetype=2) +
  theme_few()

ggsave(filename = paste0("Adult_line_w",window_size,".pdf"), plot = p, path = "./Plots", width = 20, height = 10)
  

```




### Determining Optimal Clusters

```{r}
library(factoextra)
library(NbClust)

####https://www.r-bloggers.com/2020/05/how-to-determine-the-number-of-clusters-for-k-means-in-r/

# Elbow method
fviz_nbclust(mat_all, kmeans, method = "wss") +
  geom_vline(xintercept = 2, linetype = 2)+
  labs(subtitle = "Elbow method")
# Silhouette method
fviz_nbclust(mat_all, kmeans, method = "silhouette")+
  labs(subtitle = "Silhouette method")
# Gap statistic
# nboot = 50 to keep the function speedy. 
# recommended value: nboot= 500 for your analysis.
# Use verbose = FALSE to hide computing progression.
# set.seed(123)
# fviz_nbclust(mat_all, kmeans, nstart = 25,  method = "gap_stat", nboot = 500)+
#   labs(subtitle = "Gap statistic method")

```

### Clustering using K-means

```{r}
set.seed(20)

for(k in 2:5){
  
  HM <- Heatmap(mat_all, 
      name="log2FoldChange",
      show_row_names = FALSE,
      km=k,
      col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red")),
      column_names_gp = gpar(fontsize =8)
      #heatmap_height = unit(12, "cm"), 
      #heatmap_width = unit(12, "cm")
      )

  pdf(paste0("./Plots/heatmap_","k",k,".pdf"),height=12,width=12)
  plot(HM)
  dev.off()
#   out <- data.frame()
#   ## get clusters
#   for (i in 1:length(row_order(HM))){
#     if (i == 1) {
#       clu <- t(t(row.names(mat_all[row_order(HM)[[i]],])))
#       out <- cbind(clu, paste("cluster", i, sep=""))
#       colnames(out) <- c("GeneID", "Cluster")
#     } else {
#       clu <- t(t(row.names(mat_all[row_order(HM)[[i]],])))
#       clu <- cbind(clu, paste("cluster", i, sep=""))
#       out <- rbind(out, clu)
#     }
#   }
#   
  gene_cluster <- row_order(HM)

  clu_df <- lapply(names(gene_cluster), function(i){
  out <- data.frame(coordinates = rownames(mat_all[gene_cluster[[i]],]),
                    Cluster = paste0("cluster", i), stringsAsFactors = FALSE)
  return(out)
}) %>%
  do.call(rbind, .)
  
  write.table(clu_df,paste0("kmeans_k",k,".txt"),sep="\t",quote=F,row.names=F)
  
}

```


```{r}
sessionInfo()
```
